### PR TITLE
Detect Java toolchain usage using Provider with side effect

### PR DESCRIPTION
**Pull request by alllex**
Fixes https://github.com/gradle/gradle/issues/21314

Also fixes https://github.com/gradle/gradle/issues/21368

### CODE REVIEW

**Comment by alllex** [comment]
@bot-gradle test qfl

**Comment by alllex** [comment]
@bot-gradle test this

File: subprojects/model-core/src/main/java/org/gradle/api/internal/provider/ValueSupplier.java [hunk 1]
```diff
@@ -315,10 +485,26 @@ public T orNull() {
...
+        }
+
+        @Override
+        public <R> Value<R> transform(Transformer<? extends R, ? super T> transformer) {
+            return asType();
+        }
+
+        @Override
+        public Value<T> withSideEffect(SideEffect<? super T> sideEffect) {
+            // TODO: consider if we want to extend sideEffect contract to work with missing values
```
Comment (wolfs) [inline subprojects/model-core/src/main/java/org/gradle/api/internal/provider/ValueSupplier.java:504]: Remove this todo.

File: subprojects/model-core/src/main/java/org/gradle/api/internal/provider/WithSideEffectProvider.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    @Override
+    public ProviderInternal<T> withSideEffect(SideEffect<? super T> sideEffect) {
+        return new WithSideEffectProvider<>(provider, SideEffect.composite(this.sideEffect, sideEffect));
+    }
+
+    @Override
+    public String toString() {
+        return "" + provider + " (with side effect " + sideEffect + ")";
```
Comment (wolfs) [inline subprojects/model-core/src/main/java/org/gradle/api/internal/provider/WithSideEffectProvider.java:65]: Let's figure out if using `SideEffect.toString()` makes sense here.

File: subprojects/model-core/src/test/groovy/org/gradle/api/internal/provider/WithSideEffectProviderTest.groovy [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+package org.gradle.api.internal.provider
+
+import spock.lang.Specification
+
+import java.util.concurrent.atomic.AtomicInteger
+
+
+class WithSideEffectProviderTest extends Specification {
+
+    def "runs side effect when calling '#method' on fixed value provider"() {
```
Comment (wolfs) [inline subprojects/model-core/src/test/groovy/org/gradle/api/internal/provider/WithSideEffectProviderTest.groovy:26]: Let's have this test extends `ProviderSpec`.

File: subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/JavaToolchainQueryService.java [hunk 4]
```diff
@@ -58,11 +59,17 @@ public JavaToolchainQueryService(
...
+        Transformer<T, JavaToolchain> toolFunction,
+        DefaultJavaToolchainUsageProgressDetails.JavaTool requestedTool
+    ) {
+        return findMatchingToolchain(spec)
+            .withSideEffect(toolchain -> toolchain.emitUsageEvent(requestedTool))
+            .map(toolFunction);
     }
 
-    Provider<JavaToolchain> findMatchingToolchain(JavaToolchainSpec filter) {
+    ProviderInternal<JavaToolchain> findMatchingToolchain(JavaToolchainSpec filter) {
```
Comment (wolfs) [inline subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/JavaToolchainQueryService.java:72]: Maybe add `VisibleForTesting`.

File: subprojects/model-core/src/main/java/org/gradle/api/internal/provider/ValueSupplier.java (hunk 1 above)
Comment (alllex) [reply subprojects/model-core/src/main/java/org/gradle/api/internal/provider/ValueSupplier.java]: Done

File: subprojects/model-core/src/main/java/org/gradle/api/internal/provider/WithSideEffectProvider.java (hunk 2 above)
Comment (alllex) [reply subprojects/model-core/src/main/java/org/gradle/api/internal/provider/WithSideEffectProvider.java:65]: Seems to be useful when coupled with sensible `toString` implementations for `SideEffect`:
image

File: subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/JavaToolchainQueryService.java (hunk 4 above)
Comment (alllex) [reply subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/JavaToolchainQueryService.java:74]: Done

File: subprojects/model-core/src/test/groovy/org/gradle/api/internal/provider/WithSideEffectProviderTest.groovy [hunk 5]
```diff
@@ -86,87 +99,7 @@ class WithSideEffectProviderTest extends Specification {
         "getOrElse" | { it.getOrElse(88) }
     }
 
-    def "does not run side effect when calling '#method' on missing value provider"() {
```
Comment (alllex) [inline subprojects/model-core/src/test/groovy/org/gradle/api/internal/provider/WithSideEffectProviderTest.groovy]: These removed tests are covered with identical generalized tests in the `ProviderSpec`

**Review by wolfs — APPROVED** [review]
LGTM!

**Pull request merged**